api/src/main/java/org/pmiops/workbench/cohortbuilder/FieldSetQueryBuilder.java

```diff
@@ -0,0 +1,303 @@
+        } else {
+          sqlBuilder.append("\nand\n");
+        }
+        handleColumnFilter(columnFilter, tableQueryAndConfig, sqlBuilder, paramMap);
+      }
+      sqlBuilder.append(")");
+    }
+  }
+
+  public QueryJobConfiguration buildQuery(ParticipantCriteria participantCriteria,
+      TableQueryAndConfig tableQueryAndConfig, long resultSize, long offset) {
+    TableQuery tableQuery = tableQueryAndConfig.getTableQuery();
+    TableConfig tableConfig = tableQueryAndConfig.getTableConfig();
+    List<String> columnNames = tableQuery.getColumns();
+    String tableName = tableQuery.getTableName();
+
+    StringBuilder startSql = new StringBuilder("select ");
+    startSql.append(Joiner.on(", ").join(columnNames));
+    startSql.append("\nfrom `${projectId}.${dataSetId}.");
+    startSql.append(tableName);
+    startSql.append("` ");
+    startSql.append(tableName);
+    startSql.append("\nwhere\n");
+
+    Map<String, QueryParameterValue> paramMap = new HashMap<>();
+    List<List<ColumnFilter>> columnFilters = tableQuery.getFilters();
+    if (columnFilters != null && !columnFilters.isEmpty()) {
+      if (columnFilters.size() == 1) {
+        handleColumnFilters(columnFilters.get(0), tableQueryAndConfig, startSql, paramMap);
+      } else {
+        startSql.append("(");
+        boolean first = true;
+        for (List<ColumnFilter> filterList : columnFilters) {
+          if (first) {
+            first = false;
+          } else {
+            startSql.append("\nor\n");
+          }
+          handleColumnFilters(filterList, tableQueryAndConfig, startSql, paramMap);
+        }
+        startSql.append(")");
+      }
+      startSql.append("\nand\n");
+    }
+
+    StringBuilder endSql = new StringBuilder("order by ");
+    List<String> orderBy = tableQuery.getOrderBy();
+    endSql.append(Joiner.on(", ").join(orderBy));
+    endSql.append(" limit ");
+    endSql.append(resultSize);
+    if (offset > 0) {
+      endSql.append(" offset ");
+      endSql.append(offset);
+    }
+    return participantCounter.buildQuery(participantCriteria, startSql.toString(), endSql.toString(),
+        paramMap);
+  }
+
+  public Map<String, Object> extractResults(TableQueryAndConfig tableQueryAndConfig,
+      List<FieldValue> row) {
+    TableQuery tableQuery = tableQueryAndConfig.getTableQuery();
+    TableConfig tableConfig = tableQueryAndConfig.getTableConfig();
+     List<ColumnConfig> columnConfigs = tableQuery.getColumns().stream()
+        .map(columnName -> tableQueryAndConfig.getColumn(columnName)).collect(Collectors.toList());
+    Map<String, Object> results = new HashMap<>(tableQuery.getColumns().size());
+    for (int i = 0; i < columnConfigs.size(); i++) {
+      FieldValue fieldValue = row.get(i);
+      ColumnConfig columnConfig = columnConfigs.get(i);
+      if (!fieldValue.isNull()) {
+        Object value;
+        switch (columnConfig.type) {
+          case DATE:
+            value = fieldValue.getStringValue();
```

Comment:
Will this be formatted properly, e.g. not server localized or something? Have we documented what format clients should expect to get?

Reply:
So with the code the way it is, dates are accepted as input and returned as output as yyyy-MM-dd, and datetimes as yyyy-MM-dd HH:mm:ss. There's no timezone information in there at the moment... it should be UTC for everything. We could change our datetime format to make that explicit?

Reply:
I changed it to be explicit. But now I'm wondering if we should instead just treat timestamps as longs (see #workbench-eng).

Reply:
Going to leave it as a string (with timezone) for now.

Reply:
I generally defer to the Google Cloud API style guide for these kinds of issues, which recommends using an RFC 3339 format string where possible.

https://cloud.google.com/apis/design/naming_convention#time_and_duration